### DISCUSSION TITLE
Using Hono RPC & Zod OpenAPI & Deno & React?

### DISCUSSION THREAD

**Discussion by p8952** [honojs/hono#3444] [Q&A] [answered]
I've been using Hono in Production for a while. My API Server is running on Deno using the Zod OpenAPI middleware to provide validation and generate documentation. I also have a Client which is a server side Deno App which calls my API Server.

This has been working brilliantly, I have end-to-end type safety and automatically generated API Docs 🎉

<img width="2032" alt="Screenshot 2024-09-23 at 18 14 01" src="https://github.com/user-attachments/assets/c1d980c4-58ca-4f25-9f2b-244b9adbf204">

I now need to create a second Client for my API, which is a React App, however I'm having trouble getting the Types to display in VSCode.

<img width="2032" alt="Screenshot 2024-09-23 at 18 16 22" src="https://github.com/user-attachments/assets/28e967ea-7643-4a98-8bf3-a8259e0d0eb7">

I've seen in the RPC Docs that it says:

>For the RPC types to work properly in a monorepo, in both the Client's and Server's tsconfig.json files, set `"strict": true` in `compilerOptions`.

However since I'm using Deno there is no tsconfig.json for the Server.

I've made a simplified example to show the structure of my App and Repo: https://github.com/p8952/deno-hono-react

Can anyone point me to where I'm going wrong? Thanks!

**Comment by yusukebe** [comment]
Hi @p8952 

A quick response. There's a difference between  `hono` referred from `npm:@hono/zod-openapi@0.16.0` and `hono` referred from  `hono/client`. If it's different, the type inferring does not work well. This is a TypeScript limitation.

**Comment by p8952** [comment]
Hey @yusukebe,

Thanks for the reply, when you say there's a difference, did you mean in terms of the version?

I did notice after reading your comment that the version of `hc/client` imported by my server side Deno App (where type inference works) was `=4.5.11`, while the version in my React App was `^4.6.2`.

However changing the React App version to `=4.5.11` didn't resolve my issue.

**Reply by yusukebe** [reply]
> Thanks for the reply, when you say there's a difference, did you mean in terms of the version?

This is a difficult problem. I'm not 100% sure regarding this issue, but though these are the same version, if both do not refer to the same "file," the error occurs. This is a super trouble issue.

**Comment by p8952** [comment, marked as answer]
I was able to resolve this 🎉 

Create `api/src/deno.json`:

```
{
  "lock": false,
  "imports": {
    "@hono/zod-openapi": "npm:@hono/zod-openapi@0.16.0",
    "hono/cors": "npm:hono@4.5.11/cors"
  }
}
```

Update `api/src/index.ts`:

```
import { OpenAPIHono } from "@hono/zod-openapi";
import { cors } from "hono/cors";
```

Update `api/src/routes/testRoute.ts`:

```
import { createRoute, type RouteHandler, z } from "@hono/zod-openapi";
```

Update `web/tsconfig.app.json`:

```
/* Deno Types */
"paths": {
  "@hono/zod-openapi": ["./node_modules/@hono/zod-openapi"],
  "hono/": ["./node_modules/hono/"]
}
```

Update `web/package.json`:

```
"dependencies": {
  "@hono/zod-openapi": "=0.16.0",
  "hono": "=4.5.11",
  "react": "^18.3.1",
  "react-dom": "^18.3.1"
}
```

<img width="1582" alt="Screenshot 2024-09-24 at 12 37 25" src="https://github.com/user-attachments/assets/655a2c2f-3007-44ea-8c21-7dbf34dd15a7">

I've updated my demonstration repo with the fix in case it's useful for anyone else: https://github.com/p8952/deno-hono-react
